### PR TITLE
IPA: Gitlab integration

### PR DESCRIPTION
**Pull request by velrest**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,3 +60,7 @@ make deploy
 
 ### LDAP
 To authenticate against LDAP, we used the [Passport LDAP authentication strategy](https://github.com/vesse/passport-ldapauth)
+
+If the ucs1 container exits with seg fault (error code 139), you need to either update to a never ucs-master image (and also update your config) or set `GRUB_CMDLINE_LINUX_DEFAULT="quiet vsyscall=emulate"` in `/etc/default/grub` and then do `grub-mkconfig -o /boot/grub/grub.cfg` and `reboot`.
+
+You could also use the prod LDAP server. Config is on prod custtomer-center.
```
[the remark this answers is not in the document]
Comment (velrest) [reply README.md]: :+1:

File: backend/gulpfile.js [hunk 2]
```diff
@@ -1,20 +1,28 @@
 import gulp from 'gulp'
 import eslint from 'gulp-eslint'
 import mocha from 'gulp-spawn-mocha'
+//import istanbul from 'gulp-istanbul'
 // import coverageEnforcer from 'gulp-istanbul-enforcer'
```
[the remark this answers is not in the document]
Comment (velrest) [reply backend/gulpfile.js]: :+1:

File: backend/src/gitlab/proxy.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+     *
+     */
+    proxyReqOptDecorator(proxyReqOpts) {
+      proxyReqOpts.headers['private-token'] = config.token
+      return proxyReqOpts
+    }
+  })
+}
+
+module.exports = { createProxy }
```
[the remark this answers is not in the document]
Comment (velrest) [reply backend/src/gitlab/proxy.js]: :+1:

File: backend/gulpfile.js [hunk 4]
```diff
@@ -1,20 +1,28 @@
...
       reporter: 'spec',
-      compilers: 'js:babel-register',
-      istanbul: false
+      compilers: 'js:babel-core/register'
     })
   )
+  //.pipe(istanbul.writeReports())
   //.pipe(coverageEnforcer({
   //  thresholds: {
   //    statements: 100
```
[the remark this answers is not in the document]
Comment (velrest) [reply backend/gulpfile.js]: :+1:

File: backend/gulpfile.js [hunk 5]
```diff
@@ -1,20 +1,28 @@
...
 
 let testFiles = ['tests/helper.js', 'tests/**/*-test.js']
 
+//gulp.task('pre-test', function() {
+//return gulp
+//.src(['src/**/*.js'])
+//.pipe(istanbul())
+//.pipe(istanbul.hookRequire())
+//})
+
```
[the remark this answers is not in the document]
Comment (velrest) [reply backend/gulpfile.js]: :+1:

File: frontend/app/components/gitlab-groups/component.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  projects: computed('filter', 'model', function() {
+    let search = this.get('filter.search')
+    if (search && search.trim() !== '') {
+      // Get projects which match the filter
+      return this.get('model.projects').filter(project =>
+        project.name.toLowerCase().includes(search.trim().toLowerCase())
+      )
+    } else {
+      return this.get('model.projects')
+    }
```
Comment (anehx) [inline frontend/app/components/gitlab-groups/component.js]: If else returns should not be used.. Remove the else statement to make it more readable.

File: frontend/app/gitlab/controller.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+   * @author Jonas Cosandey (jonas.cosandey@adfinis-sygroup.ch)
+   */
+  groups: computed('model', 'selection', function() {
+    if (!this.get('selection')) {
+      return this.get('model')
+    } else {
+      return [
+        this.get('model').find(group => group.id == this.get('selection.id'))
+      ]
+    }
```
Comment (anehx) [inline frontend/app/gitlab/controller.js]: Same here - no if else returns!

File: frontend/package.json [hunk 8]
```diff
@@ -28,12 +28,13 @@
     "ember-cli": "~2.18.1",
     "ember-cli-app-version": "^3.1.3",
     "ember-cli-babel": "^6.11.0",
+    "ember-cli-code-coverage": "kategengler/ember-cli-code-coverage",
```
Comment (anehx) [inline frontend/package.json]: What version, branch, tag? And why?

File: frontend/package.json [hunk 9]
```diff
@@ -52,6 +53,10 @@
     "ember-model-namespaces": "adfinis-sygroup/ember-model-namespaces",
     "ember-moment": "^7.5.0",
     "ember-notify": "^5.2.1",
+    "ember-pad": "^1.2.2",
```
Comment (anehx) [inline frontend/package.json]: Is this even used?

File: frontend/tests/acceptance/gitlab-test.js [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+module('Acceptance | GitLab', function(hooks) {
+  setupApplicationTest(hooks)
+
+  hooks.beforeEach(function() {
+    this.application = null
+  })
+
+  hooks.beforeEach(async function() {
+    this.application = startApp()
```
Comment (anehx) [inline frontend/tests/acceptance/gitlab-test.js]: Isn't this done by `setupApplicationTest`?

Comment (anehx) [reply frontend/tests/acceptance/gitlab-test.js]: And why do you have two times `hooks.beforeEach`?

File: frontend/tests/acceptance/gitlab-test.js [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+  hooks.afterEach(async function() {
+    await invalidateSession(this.application)
+    destroyApp(this.application)
+  })
+
+  test('[T08] All projects and namespaces are rendered', async function(assert) {
+    await visit('/projects')
+
+    assert.equal(find('[data-test-group]').length, 3)
+    assert.equal(find('[data-test-project]').length, 12)
```
Comment (anehx) [inline frontend/tests/acceptance/gitlab-test.js]: `assert.dom('[data-test-project]').exists({ count: 12 })` same for groups

File: frontend/tests/acceptance/gitlab-test.js [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+      .hasText('project2')
+    assert
+      .dom('[data-test-group="2"] [data-test-project-name="3"]')
+      .hasText('project3')
+    assert
+      .dom('[data-test-group="2"] [data-test-project-name="4"]')
+      .hasText('project4')
+    assert
+      .dom('[data-test-group="2"] [data-test-project-name="5"]')
+      .hasText('project5')
```
Comment (anehx) [inline frontend/tests/acceptance/gitlab-test.js]: Couldn't you solve this with a loop? This is not really what I'd call "readable"

File: frontend/tests/acceptance/gitlab-test.js [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+    assert.equal(find('[data-test-pipelines]').length, 12)
+    assert.equal(find('[data-test-commits]').length, 12)
+  })
+
+  test('[T16] Filter', async function(assert) {
+    await visit('/projects')
+    await selectChoose('[data-test-select] > *', 'test2')
+    assert.equal(find('[data-test-project]').length, 4)
+    await clearSelected('[data-test-select]')
+    assert.equal(find('[data-test-project]').length, 12)
```
Comment (anehx) [inline frontend/tests/acceptance/gitlab-test.js]: `assert.dom` ..

File: frontend/tests/acceptance/gitlab-test.js [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+        '[data-test-group="0"] [data-test-project="0"] [data-test-pipeline="2"] [data-test-stage-running]'
+      )
+      .exists()
+  })
+
+  test('[T15] Projects are displayed correct', async function(assert) {
+    await visit('/projects')
+    assert.equal(find('[data-test-project-name]').length, 12)
+    assert.equal(find('[data-test-pipelines]').length, 12)
+    assert.equal(find('[data-test-commits]').length, 12)
```
Comment (anehx) [inline frontend/tests/acceptance/gitlab-test.js]: `assert.dom` ...

File: frontend/tests/acceptance/gitlab-test.js [hunk 15]
```diff
@@ -14,10 +14,6 @@ import moment from 'moment'
 module('Acceptance | GitLab', function(hooks) {
   setupApplicationTest(hooks)
 
-  hooks.beforeEach(function() {
-    this.application = null
-  })
-
   hooks.beforeEach(async function() {
     this.application = startApp()
```
Comment (anehx) [inline frontend/tests/acceptance/gitlab-test.js]: https://github.com/emberjs/ember-qunit/blob/master/addon-test-support/ember-qunit/index.js#L142 `startApp` should not be necessary!

File: frontend/tests/acceptance/gitlab-test.js (hunk 10 above)
Comment (velrest) [reply frontend/tests/acceptance/gitlab-test.js]: no setupApplicationTest does not setup mirage in the version i have installed. the testing will be updated to the latest qunit and mirage in #71

File: docker-compose.yml [hunk 16]
```diff
@@ -1,92 +1,132 @@
...
+        driver: "none"
+      ports:
+        - '5080:80'
+        - '5443:443'
+        - '5022:22'
+
+  gitlab-runner:
+      image: 'gitlab/gitlab-runner:latest'
+      depends_on:
+        - gitlab
```
[the remark this answers is not in the document]
Comment (velrest) [reply docker-compose.yml]: :+1:

File: frontend/app/components/gitlab-groups/component.js (hunk 6 above)
Comment (velrest) [reply frontend/app/components/gitlab-groups/component.js]: :+1:

File: frontend/app/gitlab/controller.js (hunk 7 above)
Comment (velrest) [reply frontend/app/gitlab/controller.js]: :+1:

File: frontend/package.json (hunk 8 above)
Comment (velrest) [reply frontend/package.json]: :+1:

File: frontend/package.json (hunk 9 above)
Comment (velrest) [reply frontend/package.json]: :+1:

File: frontend/tests/acceptance/gitlab-test.js (hunk 11 above)
Comment (velrest) [reply frontend/tests/acceptance/gitlab-test.js]: :+1:

File: frontend/tests/acceptance/gitlab-test.js (hunk 12 above)
Comment (velrest) [reply frontend/tests/acceptance/gitlab-test.js]: :+1:

File: frontend/tests/acceptance/gitlab-test.js (hunk 13 above)
Comment (velrest) [reply frontend/tests/acceptance/gitlab-test.js]: :+1:

File: frontend/tests/acceptance/gitlab-test.js (hunk 14 above)
Comment (velrest) [reply frontend/tests/acceptance/gitlab-test.js]: :+1:

**Pull request merged**